File: frontend/src/lib/semantic-registry.ts
export const CANONICAL_SIGNAL_STATES = ['TriggeredLong', 'Watch', 'NoSetup', 'RiskOff'] as const;
export const LEGACY_SIGNAL_STATES = ['Long', 'Short', 'Side'] as const;
export const ALL_SIGNAL_STATES = [...CANONICAL_SIGNAL_STATES, ...LEGACY_SIGNAL_STATES] as const;

export type SignalState = typeof CANONICAL_SIGNAL_STATES[number];
export type LegacySignalState = typeof LEGACY_SIGNAL_STATES[number];
export type AnySignalState = typeof ALL_SIGNAL_STATES[number];

export const CANONICAL_DECISION_SEMANTICS = ['建议看多', '建议观察', '建议防守', '暂无信号'] as const;
export type DecisionSemantic = typeof CANONICAL_DECISION_SEMANTICS[number];

/**
 * Mapping between Backend Logic Tags and i18n Keys (dashboard.signal.*)
 */
export const SIGNAL_I18N_MAP = {
    '建议看多': 'triggeredLong',
    '建议观察': 'watching',
    '建议防守': 'riskOff',
    '暂无信号': 'noSignal',
    'triggeredLong': 'triggeredLong', // Support for direct canonical state
    'watching': 'watching',
    'riskOff': 'riskOff',
    'noSignal': 'noSignal',
    'Watch': 'watching',
    'RiskOff': 'riskOff',
    'TriggeredLong': 'triggeredLong',
    'NoSetup': 'noSignal',
    // Council specific mappings
    'enter': 'triggeredLong',
    'observe': 'watching',
    'defense': 'riskOff',
    'empty': 'noSignal',
    'mixed': 'mixed',
} as const;

export const ACTION_DECISIONS = ['ENTER_LONG', 'WATCH', 'DEFEND', 'NO_SIGNAL'] as const;
export const ACTION_SEMANTICS = CANONICAL_DECISION_SEMANTICS;
export type ActionDecision = typeof ACTION_DECISIONS[number];

export const DECISION_SEMANTIC_ALIASES = {
    建议进场: '建议看多',
    进场: '建议看多',
    Long: '建议看多',
    TriggeredLong: '建议看多',
    空仓: '暂无信号',
    建议空仓: '暂无信号',
    Side: '暂无信号',
    NoSetup: '暂无信号',
    防守: '建议防守',
    Short: '建议防守',
    RiskOff: '建议防守',
    观察: '建议观察',
    Watch: '建议观察',
} as const satisfies Record<string, DecisionSemantic>;

export const DECISION_ALIAS_LONG = ['建议看多', '建议进场', '进场'] as const;
export const DECISION_ALIAS_WATCH = ['建议观察', '观察'] as const;
export const DECISION_ALIAS_DEFENSE = ['建议防守', '防守'] as const;
export const DECISION_ALIAS_NO_SIGNAL = ['暂无信号', '建议空仓', '空仓'] as const;

const LEGACY_SIGNAL_SET = new Set<string>(LEGACY_SIGNAL_STATES);
const CANONICAL_SIGNAL_SET = new Set<string>(CANONICAL_SIGNAL_STATES);

export function normalizeDecisionSemantic(value: unknown, fallback: DecisionSemantic = '暂无信号'): DecisionSemantic {
    const raw = String(value ?? '').trim();
    if (!raw) return fallback;
    if ((CANONICAL_DECISION_SEMANTICS as readonly string[]).includes(raw)) return raw as DecisionSemantic;
    return DECISION_SEMANTIC_ALIASES[raw as keyof typeof DECISION_SEMANTIC_ALIASES] ?? fallback;
}

export function normalizeLayer1Status(value: unknown, fallback: SignalState = 'NoSetup'): SignalState {
    const raw = String(value ?? '').trim();
    if (CANONICAL_SIGNAL_SET.has(raw)) return raw as SignalState;
    return fallback;
}

export function normalizeOverlaySignal(value: unknown, fallback: LegacySignalState = 'Side'): LegacySignalState {
    const raw = String(value ?? '').trim();
    if (LEGACY_SIGNAL_SET.has(raw)) return raw as LegacySignalState;
    if (raw === 'TriggeredLong') return 'Long';
    if (raw === 'RiskOff') return 'Short';
    if (raw === 'Watch' || raw === 'NoSetup') return 'Side';
    return fallback;
}

export function normalizeAnySignal(value: unknown, fallback: AnySignalState = 'Side'): AnySignalState {
    const raw = String(value ?? '').trim();
    if (CANONICAL_SIGNAL_SET.has(raw)) return raw as SignalState;
    if (LEGACY_SIGNAL_SET.has(raw)) return raw as LegacySignalState;
    if (!raw) return fallback;
    const legacyFallback: LegacySignalState =
        fallback === 'TriggeredLong' ? 'Long' :
            fallback === 'RiskOff' ? 'Short' :
                'Side';
    return normalizeOverlaySignal(raw, legacyFallback);
}
/**
 * Safely resolve a database/backend value to an i18n key for the signal label.
 */
export function getSignalI18nKey(value: unknown): string {
    const normalized = normalizeDecisionSemantic(value);
    const key = SIGNAL_I18N_MAP[normalized as keyof typeof SIGNAL_I18N_MAP] ||
                SIGNAL_I18N_MAP[value as keyof typeof SIGNAL_I18N_MAP];
    return key || 'noSignal';
}

/**
 * Get i18n key for the signal's badge/description.
 */
export function getSignalBadgeI18nKey(value: unknown): string {
    const key = getSignalI18nKey(value);
    return key; // Same key used in "badge" namespace
}

File: backend/engine/semantic_registry.py
from __future__ import annotations

from typing import Literal, Optional

# Producer/Layer-1 signal semantics
CANONICAL_SIGNAL_STATES = ("TriggeredLong", "Watch", "NoSetup", "RiskOff")
LEGACY_SIGNAL_STATES = ("Long", "Short", "Side")
ALL_SIGNAL_STATES = CANONICAL_SIGNAL_STATES + LEGACY_SIGNAL_STATES
SignalState = Literal["TriggeredLong", "Watch", "NoSetup", "RiskOff"]
LegacySignalState = Literal["Long", "Short", "Side"]

# Decision/action semantics (Chinese user-facing contract)
DECISION_SEMANTIC_LONG = "建议看多"
DECISION_SEMANTIC_WATCH = "建议观察"
DECISION_SEMANTIC_DEFENSE = "建议防守"
DECISION_SEMANTIC_NO_SIGNAL = "暂无信号"

CANONICAL_DECISION_SEMANTICS = (
    DECISION_SEMANTIC_LONG,
    DECISION_SEMANTIC_WATCH,
    DECISION_SEMANTIC_DEFENSE,
    DECISION_SEMANTIC_NO_SIGNAL,
)
DecisionSemantic = Literal["建议看多", "建议观察", "建议防守", "暂无信号"]

# Keep action-layer semantics explicit even when they currently share labels.
ACTION_DECISION_IDS = ("ENTER_LONG", "WATCH", "DEFEND", "NO_SIGNAL")
ACTION_SEMANTICS = CANONICAL_DECISION_SEMANTICS

DECISION_SEMANTIC_ALIASES = {
    "建议进场": DECISION_SEMANTIC_LONG,
    "进场": DECISION_SEMANTIC_LONG,
    "Long": DECISION_SEMANTIC_LONG,
    "TriggeredLong": DECISION_SEMANTIC_LONG,
    "空仓": DECISION_SEMANTIC_NO_SIGNAL,
    "建议空仓": DECISION_SEMANTIC_NO_SIGNAL,
    "Side": DECISION_SEMANTIC_NO_SIGNAL,
    "NoSetup": DECISION_SEMANTIC_NO_SIGNAL,
    "防守": DECISION_SEMANTIC_DEFENSE,
    "Short": DECISION_SEMANTIC_DEFENSE,
    "RiskOff": DECISION_SEMANTIC_DEFENSE,
    "观察": DECISION_SEMANTIC_WATCH,
    "Watch": DECISION_SEMANTIC_WATCH,
}

DECISION_ALIAS_LONG = (DECISION_SEMANTIC_LONG, "建议进场", "进场")
DECISION_ALIAS_WATCH = (DECISION_SEMANTIC_WATCH, "观察")
DECISION_ALIAS_DEFENSE = (DECISION_SEMANTIC_DEFENSE, "防守")
DECISION_ALIAS_NO_SIGNAL = (DECISION_SEMANTIC_NO_SIGNAL, "建议空仓", "空仓")

ACTION_DECISION_BY_SEMANTIC = {
    DECISION_SEMANTIC_LONG: "ENTER_LONG",
    DECISION_SEMANTIC_WATCH: "WATCH",
    DECISION_SEMANTIC_DEFENSE: "DEFEND",
    DECISION_SEMANTIC_NO_SIGNAL: "NO_SIGNAL",
}


def normalize_decision_semantic(
    value: object,
    default: str = DECISION_SEMANTIC_NO_SIGNAL,
) -> str:
    raw = str(value or "").strip()
    if not raw:
        return default
    if raw in CANONICAL_DECISION_SEMANTICS:
        return raw
    return DECISION_SEMANTIC_ALIASES.get(raw, default)


def is_canonical_decision_semantic(value: object) -> bool:
    return normalize_decision_semantic(value, "") in CANONICAL_DECISION_SEMANTICS


def semantic_from_layer1(
    layer1_status: Optional[str],
    signal: Optional[str],
) -> DecisionSemantic:
    if layer1_status == "TriggeredLong":
        return DECISION_SEMANTIC_LONG
    if layer1_status == "Watch":
        return DECISION_SEMANTIC_WATCH
    if layer1_status == "RiskOff":
        return DECISION_SEMANTIC_DEFENSE
    if layer1_status == "NoSetup":
        return DECISION_SEMANTIC_NO_SIGNAL
    if signal == "Long":
        return DECISION_SEMANTIC_LONG
    return DECISION_SEMANTIC_WATCH


def to_action_decision_id(decision_semantic: object) -> str:
    semantic = normalize_decision_semantic(decision_semantic, DECISION_SEMANTIC_NO_SIGNAL)
    return ACTION_DECISION_BY_SEMANTIC.get(semantic, "NO_SIGNAL")
